tests/unit/s2n_client_hello_get_supported_groups_test.c

```diff
@@ -0,0 +1,390 @@
+
+            uint16_t one_and_a_half_groups_size = 3;
+            EXPECT_SUCCESS(s2n_stuffer_write_uint16(&extension_stuffer, one_and_a_half_groups_size));
+
+            uint16_t supported_groups[S2N_TEST_SUPPORTED_GROUPS_LIST_COUNT] = { 0 };
+            uint16_t supported_groups_count = 0;
+            int ret = s2n_client_hello_get_supported_groups(&client_hello, supported_groups,
+                    s2n_array_len(supported_groups), &supported_groups_count);
+            EXPECT_FAILURE_WITH_ERRNO(ret, S2N_ERR_INVALID_PARSED_EXTENSIONS);
+
+            EXPECT_EQUAL(supported_groups_count, 0);
+        }
+    }
+
+    /* Ensure that the supported groups in the client hello are written to the output array. */
+    {
+        struct s2n_client_hello client_hello = { 0 };
+
+        s2n_extension_type_id supported_groups_id = 0;
+        EXPECT_SUCCESS(s2n_extension_supported_iana_value_to_id(S2N_EXTENSION_SUPPORTED_GROUPS, &supported_groups_id));
+
+        s2n_parsed_extension *supported_groups_extension = &client_hello.extensions.parsed_extensions[supported_groups_id];
+        supported_groups_extension->extension_type = S2N_EXTENSION_SUPPORTED_GROUPS;
+
+        for (uint16_t test_groups_count = 0; test_groups_count < S2N_TEST_SUPPORTED_GROUPS_LIST_COUNT; test_groups_count++) {
+            uint16_t test_groups_list_size = test_groups_count * 2;
+
+            uint8_t test_groups_list_data[S2N_TEST_SUPPORTED_GROUPS_LIST_SIZE] = { 0 };
+            struct s2n_blob test_groups_list_blob = { 0 };
+            EXPECT_SUCCESS(s2n_blob_init(&test_groups_list_blob, test_groups_list_data, test_groups_list_size));
+            EXPECT_OK(s2n_get_public_random_data(&test_groups_list_blob));
+
+            uint8_t extension_data[S2N_TEST_SUPPORTED_GROUPS_EXTENSION_SIZE] = { 0 };
+            struct s2n_blob extension_blob = { 0 };
+            EXPECT_SUCCESS(s2n_blob_init(&extension_blob, extension_data, sizeof(extension_data)));
+            supported_groups_extension->extension = extension_blob;
+
+            struct s2n_stuffer extension_stuffer = { 0 };
+            EXPECT_SUCCESS(s2n_stuffer_init(&extension_stuffer, &extension_blob));
+            EXPECT_SUCCESS(s2n_stuffer_write_uint16(&extension_stuffer, test_groups_list_size));
+            EXPECT_SUCCESS(s2n_stuffer_write(&extension_stuffer, &test_groups_list_blob));
+
+            uint16_t supported_groups[S2N_TEST_SUPPORTED_GROUPS_LIST_COUNT] = { 0 };
+            uint16_t supported_groups_count = 0;
+            EXPECT_SUCCESS(s2n_client_hello_get_supported_groups(&client_hello, supported_groups,
+                    s2n_array_len(supported_groups), &supported_groups_count));
+            EXPECT_EQUAL(supported_groups_count, test_groups_count);
+
+            struct s2n_stuffer test_groups_list_stuffer = { 0 };
+            EXPECT_SUCCESS(s2n_stuffer_init_written(&test_groups_list_stuffer, &test_groups_list_blob));
+
+            for (size_t i = 0; i < supported_groups_count; i++) {
```

Comment:
If you're just comparing two arrays, can you use EXPECT_BYTEARRAY_EQUAL?

Reply:
Unfortunately I'm not sure it's possible here, since the raw supported groups from the client has network-order endianness, and the parsed supported groups returned by `s2n_client_hello_get_supported_groups` has the machine's endianness. It's a similar problem to https://github.com/aws/s2n-tls/pull/4216#discussion_r1333230944.

I'll add a comment here too, though.